### PR TITLE
Dev/qp solv integration

### PR DESCRIPTION
**Pull request by ashishrao7**
Issue Number: https://github.com/intel-innersource/frameworks.ai.lava.userstories/issues/220
https://github.com/intel-innersource/frameworks.ai.lava.userstories/issues/211

Objective of pull request: Enable QP solver to be called from lava-optimization and add new tutorial.

## Pull request checklist

Your PR fulfills the following requirements:
- [x] [Issue](https://github.com/lava-nc/lava-optimization/issues) created that explains the change and why it's needed
- [x] Tests are part of the PR (for bug fixes / features)
- [x ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`pyb`) passes locally
- [x] Build tests (`pyb -E unit`) or (`python -m unittest`) passes locally

## Pull request type

 

 

Please check your PR type:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?

- The QP solver cannot be called from the generic qp solver interface

## What is the new behavior?

- The QP solver can be called with both Loihi and CPU backends by the user. Floating point models supported for now.

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

**Comment by ashishrao7** [comment]
Opened a PR to initiate the process of merging. Things that still need to be done apart from merging changes by @phstratmann  and  @AlessandroPierro 
- moving qp under generic and corresponding refactoring
- linting
- Documentation 
Tutorial will be added in another PR

**Comment by ashishrao7** [comment]
Added new QP tutorial (Tutorial 1)

File: src/lava/lib/optimization/problems/problems.py [hunk 1]
```diff
@@ -71,16 +72,16 @@ def __init__(self, q: npt.ArrayLike):
...
+        self._variables._discrete = DiscreteVariables(domains=[2] * q.shape[0])
 
     @property
     def variables(self):
         """Binary variables of the QUBO problem."""
-        return self._b_variables
+        return self._variables
 
     @property
     def num_variables(self):
```
Comment (phstratmann) [inline src/lava/lib/optimization/problems/problems.py:83]: from num_variables, I would assume to find the total number of variables, not just the discrete variables. Should be fine to implement it as such, should not break any previous code in the INRC since for the QUBO solver there are only discrete variables.

File: src/lava/lib/optimization/problems/problems.py [hunk 2]
```diff
@@ -177,124 +180,156 @@ def verify_solution(self, solution):
...
-        ):
-            raise ValueError(
-                "Please properly define your Equality constraints. Supply \
-                all A_eq and k_eq."
-            )
+        super().__init__()
+        self._variables._continuous = ContinuousVariables(
+            num_variables=hessian.shape[0]
+        )
+        self.q_cost = Cost(linear_offset, hessian)
```
Comment (phstratmann) [inline src/lava/lib/optimization/problems/problems.py:226]: Not sure if you want to add a getter/setter for q_cost, and use self._q_cost instead.
Up to you.

File: src/lava/lib/optimization/problems/problems.py [hunk 3]
```diff
@@ -177,124 +180,156 @@ def verify_solution(self, solution):
...
-            self._constraint_biases = None
+    @property
+    def constraints(self):
+        return self._constraints.arithmetic
 
-        if constraint_hyperplanes_eq is not None:
-            self._constraint_hyperplanes_eq = constraint_hyperplanes_eq
-            self._constraint_biases_eq = constraint_biases_eq
+    @property
+    def hessian(self) -> np.ndarray:
```
Comment (phstratmann) [inline src/lava/lib/optimization/problems/problems.py:249]: Be consistent:
npt.ArrayLike?

Comment (phstratmann) [reply src/lava/lib/optimization/problems/problems.py:249]: Also at other occassions.

File: src/lava/lib/optimization/problems/variables.py [hunk 4]
```diff
@@ -93,7 +94,7 @@ def variable_set(self):
     @property
     def num_variables(self):
         """Number of variables in this set."""
-        return len(self._bounds)
+        return self._num_variables
```
Comment (phstratmann) [inline src/lava/lib/optimization/problems/variables.py:97]: Why not 
`return len(self.domains) if self.domains is not None else None`
like in line 69?

File: src/lava/lib/optimization/solvers/generic/builder.py [hunk 5]
```diff
@@ -62,6 +62,7 @@ def verify_model_exists(self):
     def create_solver_process(
         self,
         problem: OptimizationProblem,
+        backend,
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/builder.py:65]: For typing:
Is backend a 'AbstractComputeResource' or a 'str'?

Comment (phstratmann) [reply src/lava/lib/optimization/solvers/generic/builder.py:65]: Als below

File: src/lava/lib/optimization/solvers/generic/builder.py [hunk 6]
```diff
@@ -140,55 +142,73 @@ def _create_process_constructor(
...
+                and problem.variables.discrete.num_variables is not None
             ):
                 self.discrete_variables = Var(
                     shape=(
-                        problem.variables.num_variables,
+                        problem.variables.discrete.num_variables,
                         # problem.variables.domain_sizes[0]
                     )
                 )
+                self.is_discrete = 1
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/builder.py:183]: Hm... If some variables are discrete and some are not, self.is_discrete would be 1. But would that correctly represent whether it's a concrete solver?

File: src/lava/lib/optimization/solvers/generic/builder.py [hunk 7]
```diff
@@ -140,55 +142,73 @@ def _create_process_constructor(
...
-                shape=(problem.variables.num_variables,)
-            )
-            # Total cost = optimality_first_byte << 24 + optimality_last_bytes
-            self.optimality_last_bytes = Var(shape=(1,))
-            self.optimality_first_byte = Var(shape=(1,))
-            self.optimum = Var(shape=(2,))
-            self.feasibility = Var(shape=(1,))
-            self.solution_step = Var(shape=(1,))
-            self.cost_monitor = Var(shape=(1,))
+                if self.is_discrete:
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/builder.py:188]: Does is_discrete=1 imply that the cost is quadratic? It does currently (QUBO), but that is unlikely to be general.

File: src/lava/lib/optimization/problems/problems.py (hunk 2 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/problems/problems.py:226]: private variables shouldn't be accessed like that. Not good practice

File: src/lava/lib/optimization/problems/problems.py [hunk 8]
```diff
@@ -177,124 +180,156 @@ def verify_solution(self, solution):
...
+        return self.constraint_hyperplanes_eq @ sol - self.constraint_biases_eq
+
+    def precondition_problem(self, iterations=5, type="ruiz"):
+        if type == "ruiz":
+            self._ruiz_precondition(iterations=iterations)
+        else:
+            raise NotImplementedError(
+                "Only Ruiz preconditioning is enabled \
+                                      for QP problems at the moment"
+            )
```
[the remark this answers is not in the document]
Comment (ashishrao7) [reply src/lava/lib/optimization/problems/problems.py:294]: This is subjective. It's probably generated by a linter. Not practical to make every string look the same. As long as it satisfies flakeheaven it should be fine

File: src/lava/lib/optimization/solvers/generic/builder.py (hunk 7 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/builder.py:188]: image
is_discrete is set when the problem has discrete variables. 
Keep in mind that the code has to be refactored as more solvers are added in the future. Making it too generic is a huge effort in terms of planning and software engineering for which I don't have the bandwidth atm.

Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/builder.py:188]: will remove the ghost code in an upcoming commit

File: src/lava/lib/optimization/solvers/generic/builder.py (hunk 6 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/builder.py:183]: we tackle mixed problems when we have a mixed problems solver

File: src/lava/lib/optimization/solvers/generic/builder.py (hunk 5 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/builder.py:65]: Okay adding type hints in an upcoming commit

File: src/lava/lib/optimization/solvers/generic/solver.py [hunk 9]
```diff
@@ -61,6 +71,11 @@
     from lava.lib.optimization.solvers.generic.read_gate.ncmodels import (
         ReadGateCModel,
     )
+
+    from lava.lib.optimization.solvers.generic.qp.ncmodels import (
```
Comment (GaboFGuerra) [inline src/lava/lib/optimization/solvers/generic/solver.py:75]: why does the name specify L2? We cannot implement this models on L1 anyway.

File: src/lava/lib/optimization/solvers/generic/solver.py [hunk 10]
```diff
@@ -405,6 +431,7 @@ def _get_run_config(
         )
 
         if backend in CPUS:
+            # just add CPU processes here in the dictionary for QP
```
Comment (GaboFGuerra) [inline src/lava/lib/optimization/solvers/generic/solver.py:434]: Please delete the comment above, it seems to be a leftover from the development syncs.

File: src/lava/lib/optimization/solvers/generic/solver.py [hunk 11]
```diff
@@ -414,11 +441,14 @@ def _get_run_config(
...
                 NEBM: NEBMPyModel,
                 QuboScif: PyModelQuboScifFixed,
+                ProportionalIntegralNeuronsPIPGeq: PyPIneurPIPGeqModel,
+                ProjectedGradientNeuronsPIPGeq: PyProjGradPIPGeqModel,
             }
             return Loihi1SimCfg(
                 exception_proc_model_map=pdict, select_sub_proc_model=True
             )
         elif backend in NEUROCORES:
+            # just add neurocore processes here in the dictionary for QP
```
Comment (GaboFGuerra) [inline src/lava/lib/optimization/solvers/generic/solver.py:451]: Same with this comment above

File: src/lava/lib/optimization/solvers/generic/sub_process_models.py [hunk 12]
```diff
@@ -9,22 +9,218 @@
...
+class ContinuousVariablesModel(AbstractSubProcessModel):
+    """Model for the ContinuousVariables process."""
+
+    def __init__(self, proc):
+        # Instantiate child processes
+        # The input shape is a 2D vector (shape of the weight matrix).
+        backend = proc.backend
+        neuron_model = proc.hyperparameters.get("neuron_model", "qp/lp-pipg")
+
+        if neuron_model == "qp/lp-pipg":
```
Comment (GaboFGuerra) [inline src/lava/lib/optimization/solvers/generic/sub_process_models.py:73]: Comments in the following lines don't seem intended for release.

File: src/lava/lib/optimization/solvers/generic/sub_process_models.py [hunk 13]
```diff
@@ -9,22 +9,218 @@
...
+            # adding them here to show that they are need for the neurons models
+            # since some values are calculated based on these weights
```
Comment (GaboFGuerra) [inline src/lava/lib/optimization/solvers/generic/sub_process_models.py:75]: change the upper-case to lower-case on the names on the next two lines

File: src/lava/lib/optimization/solvers/generic/sub_process_models.py [hunk 14]
```diff
@@ -9,22 +9,218 @@
...
+@requires(CPU)
+class ContinuousVariablesModel(AbstractSubProcessModel):
+    """Model for the ContinuousVariables process."""
+
+    def __init__(self, proc):
+        # Instantiate child processes
+        # The input shape is a 2D vector (shape of the weight matrix).
+        backend = proc.backend
+        neuron_model = proc.hyperparameters.get("neuron_model", "qp/lp-pipg")
+
```
Comment (GaboFGuerra) [inline src/lava/lib/optimization/solvers/generic/sub_process_models.py:72]: is the following conditional still relevant after integration and after deprecation of GDCC?

File: src/lava/lib/optimization/solvers/generic/sub_process_models.py [hunk 15]
```diff
@@ -9,22 +9,218 @@
...
+            proc.in_ports.a_in.connect(self.conn_A.s_in)
+            self.conn_A.a_out.connect(self.ProInt.a_in)
+            self.ProInt.s_out.connect(self.conn_A_T.s_in)
+            self.conn_A_T.a_out.connect(proc.out_ports.s_out)
+            proc.vars.constraint_assignment.alias(
+                self.ProInt.constraint_neuron_state
+            )
+        else:
+            AssertionError("Unknown neuron model specified")
```
Comment (GaboFGuerra) [inline src/lava/lib/optimization/solvers/generic/sub_process_models.py:225]: Some changes below seem unrelated to QP integration or linting, is this intended?

File: src/lava/lib/optimization/utils/qp_processing.py
Comment (GaboFGuerra) [inline src/lava/lib/optimization/utils/qp_processing.py:1]: "qp_processing" does not seem the most appropriate, e.g., this function does not seem to be restricted to QP, although it's used for it, it might be used in other places.
maybe utils/float_to_fix_point.py utils/datatype_converter.py,  or something like that.

File: src/lava/lib/optimization/utils/qp_processing.py [hunk 16]
```diff
@@ -0,0 +1,27 @@
+# Copyright (C) 2022 Intel Corporation
+# SPDX-License-Identifier: BSD-3-Clause
+# See: https://spdx.org/licenses/
+import numpy as np
+
+
```
Comment (GaboFGuerra) [inline src/lava/lib/optimization/utils/qp_processing.py:6]: I think mat is not trivially understood as matrix, I would use full spelling

Comment (GaboFGuerra) [reply src/lava/lib/optimization/utils/qp_processing.py:6]: similarly with man_bits

File: tests/lava/lib/optimization/solvers/generic/test_solver_cpu_backend.py [hunk 17]
```diff
@@ -1,32 +1,31 @@
...
 
 
-class TestOptimizationSolver(unittest.TestCase):
+class TestOptimizationSolverQUBO(unittest.TestCase):
     def setUp(self) -> None:
         self.problem = QUBO(
-            np.array([[-5, 2, 4, 0],
-                      [2, -3, 1, 0],
-                      [4, 1, -8, 5],
-                      [0, 0, 5, -6]]))
```
Comment (GaboFGuerra) [inline tests/lava/lib/optimization/solvers/generic/test_solver_cpu_backend.py:29]: If possible leave the matrix style instead of the nested list looking one

File: tests/lava/lib/optimization/solvers/generic/test_solver_cpu_backend.py [hunk 18]
```diff
@@ -131,19 +131,112 @@ def test_state_tracking(self):
         )
 
 
-def solve_workload(problem, reference_solution, noise_precision=5,
-                   noise_amplitude=1, on_tau=-3):
+class TestOptimizationSolverQP(unittest.TestCase):
+    def setUp(self) -> None:
+        root = os.path.dirname(os.path.abspath(__file__))
+        qp_data = np.load(root + "/data/qp/ex_qp_small.npz")
```
Comment (GaboFGuerra) [inline tests/lava/lib/optimization/solvers/generic/test_solver_cpu_backend.py:137]: I would suggest not using upper case for variable names

**Review by GaboFGuerra — APPROVED** [review]
Thanks Ashish, great work!

Left a few comments but am approving as nothing is major and can be addressed on a follow up release. Please read the comments and decide which to address.

In particular, please check the changes to sub_process_model.py don't include any unintended changes on the QUBO side.

File: src/lava/lib/optimization/problems/problems.py [hunk 19]
```diff
@@ -19,6 +19,7 @@
     ContinuousVariables,
     DiscreteVariables,
 )
+from scipy import sparse
```
[the remark this answers is not in the document]
Comment (phstratmann) [reply src/lava/lib/optimization/problems/problems.py:22]: I checked, it is (see poetry.lock)

File: src/lava/lib/optimization/solvers/generic/builder.py (hunk 7 above)
Comment (phstratmann) [reply src/lava/lib/optimization/solvers/generic/builder.py:188]: Just noticed that you also use the self.is_continuous variable.
Maybe worth renaming them to
self.has_continuous_variables
self.has_continuous_variables?

File: src/lava/lib/optimization/solvers/generic/dataclasses.py [hunk 20]
```diff
@@ -73,6 +93,10 @@ def local_cost(self):
     def variables_assignment(self):
         return self.discrete.variable_assignment
 
+    @property
+    def variables_assignment_cont(self):
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/dataclasses.py:97]: What does _cont refer to? For me, this looks more like a def variable_assignmende than the actual def variable_assignment above.

File: src/lava/lib/optimization/solvers/generic/hierarchical_processes.py [hunk 21]
```diff
@@ -15,9 +16,30 @@ class ContinuousVariablesProcess(AbstractProcess):
...
+    ) -> None:
+        super().__init__(
+            shape=shape,
+            name=name,
+            problem=problem,
+            log_config=log_config,
+        )
+
+        self.num_variables = np.prod(shape)
+        self.backend = backend
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/hierarchical_processes.py:36]: Does the Process still need knwoledge of its backend? The idea of the Process was that it is backend agnostic.

File: src/lava/lib/optimization/solvers/generic/qp/processes.py [hunk 22]
```diff
@@ -304,7 +304,7 @@ class ProjectedGradientNeuronsPIPGeq(AbstractProcess):
     process with this solver. Use the standard Lava Dense process.
     Intialize the ProjectedGradientNeuronsPIPGeq process.
     Implements the abstract behaviour
-    qp_neuron_state -= alpha*(a_in_qc + grad_bias + a_in_cn)
+    qp_neuron_state -= alpha*(a_in + grad_bias)
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/qp/processes.py:307]: Could you elaborate why you used a_in_cn previously, but is not needed anymore? Will this change the behavior related to the model used for the NeurIPS paper?

File: src/lava/lib/optimization/solvers/generic/solution_finder/models.py [hunk 23]
```diff
@@ -47,56 +71,126 @@ def __init__(self, proc):
...
+                    Q_exp_new = -correction_exp + Q_pre_fp_exp
+                    self.cost_minimizer = CostMinimizer(
+                        Sparse(
+                            weights=csr_matrix(Q_pre_fp_man),
+                            weight_exp=Q_exp_new,
+                            num_message_bits=24,
+                        )
+                    )
+                else:
+                    raise NotImplementedError(str(backend) + BACKEND_MSG)
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/solution_finder/models.py:166]: Didn't know about BACKEND_MSG, elegant way. Great.

File: src/lava/lib/optimization/solvers/generic/solution_finder/process.py [hunk 24]
```diff
@@ -15,23 +15,29 @@ def __init__(
         cost_diagonal,
         cost_coefficients,
         constraints,
+        backend,
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/solution_finder/process.py:18]: Also here: Is there any way to make the Process truly backend agnostic?

File: src/lava/lib/optimization/solvers/generic/solution_finder/process.py [hunk 25]
```diff
@@ -15,23 +15,29 @@ def __init__(
...
             hyperparameters=hyperparameters,
             discrete_var_shape=discrete_var_shape,
             continuous_var_shape=continuous_var_shape,
+            problem=problem,
             name=name,
             log_config=log_config,
         )
-        self.variables_assignment = Var(shape=discrete_var_shape, init=(1,))
+        self.variables_assignment = Var(
+            shape=discrete_var_shape or continuous_var_shape, init=(1,)
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/solution_finder/process.py:39]: Hm... For the future, could we use discrete_var_shape + continuous_var_shape here (of course accounting for None values)?

File: src/lava/lib/optimization/solvers/generic/solver.py [hunk 26]
```diff
@@ -450,20 +482,56 @@ def _prepare_profiler(self, config: SolverConfig, run_cfg) -> None:
...
+            best_cost = SolutionReadoutPyModel.decode_cost(best_cost)
+            best_timestep = (
+                self.solver_process.solution_step.aliased_var.get() - 2
+            )
+        best_state = self._get_best_state(config, int(idx))
+
+        if self.solver_process.is_discrete:
+            return best_state, int(best_cost), int(best_timestep)
+        else:
+            return best_state, None, None
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/solver.py:497]: Why do continuous problems not return the best cost and best time step?

File: src/lava/lib/optimization/solvers/generic/solver.py [hunk 27]
```diff
@@ -450,20 +482,56 @@ def _prepare_profiler(self, config: SolverConfig, run_cfg) -> None:
...
 
     def _get_best_state(self, config: SolverConfig, idx: int):
+        if self._is_problem_discrete():
+            discrete_values = self._get_and_decode_discrete_vars(config, idx)
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/solver.py:501]: DIdn't you specify the variables self.continuous and self.discrete at some point? Could they be used here?

File: src/lava/lib/optimization/solvers/generic/solver.py [hunk 28]
```diff
@@ -450,20 +482,56 @@ def _prepare_profiler(self, config: SolverConfig, run_cfg) -> None:
...
-            idx = int(idx)
             raw_solution = np.asarray(
-                self.solver_process.finders[idx].variables_assignment.get()
-            ).astype(np.int32)
+                self.solver_process.finders[idx]
+                .variables_assignment.get()
+                .astype(np.int32)
+            )
             raw_solution &= 0x3F
             return raw_solution.astype(np.int8) >> 5
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/solver.py:522]: Is this actually working for non-binary discrete value?

File: src/lava/lib/optimization/solvers/generic/sub_process_models.py [hunk 29]
```diff
@@ -52,65 +248,80 @@ def __init__(self, proc):
         if neuron_model == "nebm":
             temperature = proc.hyperparameters.get("temperature", 1)
             refract = proc.hyperparameters.get("refract", 0)
-            init_value = proc.hyperparameters.get("init_value",
```
[the remark this answers is not in the document]
Comment (phstratmann) [reply src/lava/lib/optimization/solvers/generic/sub_process_models.py:55]: In general, I would suggest not to touch the NEBM/NEBM-SA models for QP integration.

File: tests/lava/lib/optimization/solvers/generic/test_solver_cpu_backend.py [hunk 30]
```diff
@@ -94,34 +95,33 @@ def test_subprocesses_connections(self):
         )
 
     def test_qubo_cost_defines_weights(self):
+        from lava.lib.optimization.solvers.generic.solution_finder.models\
```
Comment (phstratmann) [inline tests/lava/lib/optimization/solvers/generic/test_solver_cpu_backend.py:98]: Is there a reason to import this here, instead at the top of the file?

**Review by phstratmann — CHANGES_REQUESTED** [review]
Wow, that is a tremendous amount of great work! Thanks for that.
A few comments throughout. But my main comment: For the future, is it possible to provide such work in smaller chunks? Seems dangerous to PR such a large amount of work two days before a planned release.

**Review by phstratmann — APPROVED** [review]

File: src/lava/lib/optimization/solvers/generic/solver.py (hunk 28 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/solver.py:522]: It works for now. Non-binary discrete values are not in the library. Refactor when the time arises.

File: src/lava/lib/optimization/solvers/generic/solution_finder/process.py (hunk 24 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/solution_finder/process.py:18]: no

File: src/lava/lib/optimization/solvers/generic/qp/processes.py (hunk 22 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/qp/processes.py:307]: no it won't. I optimized pymodel to use only 1 inport and 1 outport like the ncmodel

File: src/lava/lib/optimization/solvers/generic/hierarchical_processes.py (hunk 21 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/hierarchical_processes.py:36]: it does

File: src/lava/lib/optimization/problems/variables.py (hunk 4 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/problems/variables.py:97]: Because the data you get in the interface is different for continuous time and discrete problems. I didn't make any changes to the discrete interface

File: src/lava/lib/optimization/problems/problems.py (hunk 1 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/problems/problems.py:83]: what do you mean? It's a QUBO problem. It has only binary variables by definition

File: src/lava/lib/optimization/problems/problems.py (hunk 3 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/problems/problems.py:249]: done

File: src/lava/lib/optimization/solvers/generic/sub_process_models.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/sub_process_models.py:55]: They are mainly line number changes. I cannot make integration changes without adding code in this file for the QP solver. I didn't touch the logic of the QUBO stuff. All the tests pass

File: src/lava/lib/optimization/solvers/generic/solution_finder/process.py (hunk 25 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/solution_finder/process.py:39]: We could in the future

File: src/lava/lib/optimization/solvers/generic/dataclasses.py (hunk 20 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/dataclasses.py:97]: continous

File: src/lava/lib/optimization/solvers/generic/dataclasses.py [hunk 31]
```diff
@@ -37,9 +37,21 @@ class ConstraintEnforcing:
     """Processes implementing an optimization problem's constraints and their
     enforcing."""
 
-    continuous: ContinuousConstraintsProcess
-    discrete: DiscreteConstraintsProcess
-    mixed: MixedConstraintsProcess
+    continuous: ContinuousConstraintsProcess = None
+    # discrete: DiscreteConstraintsProcess
+    # mixed: MixedConstraintsProcess
```
[the remark this answers is not in the document]
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/dataclasses.py:42]: uncommented in upcoming commit

**Comment by ashishrao7** [comment]
Thanks for the reviews. I had difficulty seeing the code that you were commenting for in the preview pane. Maybe you should calibrate the positioning of your comments next time so that I can see the problem in the preview pane itself. This saves time in opening each file where there's a comment and looking at the issue.

File: src/lava/lib/optimization/solvers/generic/solver.py (hunk 26 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/solver.py:497]: because the QP solver doesn't use the Solution Reader and read gate yet

File: tests/lava/lib/optimization/solvers/generic/test_solver_cpu_backend.py (hunk 18 above)
Comment (ashishrao7) [reply tests/lava/lib/optimization/solvers/generic/test_solver_cpu_backend.py:137]: I also don't like that practice but in this case for QP it makes life easier and popular solvers also do it.

File: tests/lava/lib/optimization/solvers/generic/test_solver_cpu_backend.py (hunk 30 above)
Comment (ashishrao7) [reply tests/lava/lib/optimization/solvers/generic/test_solver_cpu_backend.py:98]: resolved in upcoming commit

File: src/lava/lib/optimization/solvers/generic/builder.py (hunk 7 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/builder.py:188]: made change

File: src/lava/lib/optimization/problems/problems.py [hunk 32]
```diff
@@ -177,124 +180,156 @@ def verify_solution(self, solution):
         raise NotImplementedError
 
 
-class QP:
+class QP(OptimizationProblem):
     """A Rudimentary interface for the QP solver. Inequality Constraints
     should be of the form Ax<=k. Equality constraints are expressed as
     sandwiched inequality constraints. The cost of the QP is of the form
     1/2*x^t*Q*x + p^Tx
```
[the remark this answers is not in the document]
Comment (ashishrao7) [reply src/lava/lib/optimization/problems/problems.py:187]: deleted

File: src/lava/lib/optimization/solvers/generic/dataclasses.py [hunk 33]
```diff
@@ -37,9 +37,21 @@ class ConstraintEnforcing:
...
+
+    @property
+    def state_in(self):
+        return self.continuous.a_in
+
+    @property
+    def state_out(self):
+        return self.continuous.s_out
+
+    @property
```
[the remark this answers is not in the document]
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/dataclasses.py:52]: They are the variables of the ConstraintEnforcing class

File: src/lava/lib/optimization/solvers/generic/solution_finder/models.py [hunk 34]
```diff
@@ -47,56 +71,126 @@ def __init__(self, proc):
...
-        )
+            self.constraints.continuous = ContinuousConstraintsProcess(
+                shape_in=continuous_var_shape,
+                shape_out=continuous_var_shape,
+                backend=backend,
+                hyperparameters=hyperparameters,
+                problem=problem,
+            )
+            self.cost_minimizer = None
+            Q_pre = problem.hessian
```
[the remark this answers is not in the document]
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/solution_finder/models.py:141]: I am not in support of having lower case for QP variables. It makes them look like vectors. standard solvers also use upper case
image

File: src/lava/lib/optimization/solvers/generic/solution_finder/models.py [hunk 35]
```diff
@@ -47,56 +71,126 @@ def __init__(self, proc):
...
+            if cost_coefficients is not None:
+                if backend in CPUS:
+                    self.cost_minimizer = CostMinimizer(
+                        Sparse(
+                            weights=csr_matrix(Q_pre),
+                            num_message_bits=64,
+                        )
+                    )
+                elif backend in NEUROCORES:
```
[the remark this answers is not in the document]
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/solution_finder/models.py:150]: Keeping upper case for reasons mentioned above

File: src/lava/lib/optimization/solvers/generic/solver.py (hunk 9 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/solver.py:75]: no, but maybe we can in L3

File: src/lava/lib/optimization/solvers/generic/sub_process_models.py (hunk 12 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/sub_process_models.py:73]: They are documentation for posterity. They were intentionally left there.

File: src/lava/lib/optimization/solvers/generic/sub_process_models.py (hunk 14 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/sub_process_models.py:72]: Leaving as upper case for reasons mentioned before

File: src/lava/lib/optimization/solvers/generic/sub_process_models.py (hunk 15 above)
Comment (ashishrao7) [reply src/lava/lib/optimization/solvers/generic/sub_process_models.py:225]: They were moved below because of addition of code above. I have not touched the logic of QUBO

File: tests/lava/lib/optimization/solvers/generic/test_solver_cpu_backend.py (hunk 17 above)
Comment (ashishrao7) [reply tests/lava/lib/optimization/solvers/generic/test_solver_cpu_backend.py:29]: this is generated by the linting tool. Impractical to manually correct all matrices in text cases by hand.
